loader/loader.go

```diff
@@ -1,5 +1,203 @@
 package loader
 
-// Loader represents loading process from a bitcoin client.
-type Loader interface {
+import (
+	"context"
+	"sync"
+
+	"github.com/IlliniBlockchain/etl-bitcoin/client"
+	"github.com/IlliniBlockchain/etl-bitcoin/database"
+	"github.com/IlliniBlockchain/etl-bitcoin/types"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"golang.org/x/sync/errgroup"
+)
+
+// ILoaderManager outlines an interface for a loader manager.
+type ILoaderManager interface {
+	SendInput(BlockRange, database.DBTx)
+	Close()
+}
+
+// LoaderManager stores state for managing loaders for loading data from a `Client` to
+// a `Database`.
+type LoaderManager struct {
+	client  client.Client
+	db      database.Database
+	inputCh chan *LoaderMsg[BlockRange]
+
+	ctx      context.Context
+	stopOnce sync.Once
+	g        *errgroup.Group
+}
+
+// LoaderMsg stores state for data being passed through loaders.
+type LoaderMsg[T any] struct {
+	blockRange BlockRange
+	dbTx       database.DBTx
+	data       T
+}
+
+// LoaderOptions represents a map of arbitrary options when constructing a LoaderManager.
+type LoaderOptions map[string]interface{}
+
+type BlockRange struct {
+	startBlockHeight int64
+	endBlockHeight   int64
+}
+
+// NewLoaderManager creates a new LoaderManager and initiates goroutines for the loaders in a pipeline.
+func NewLoaderManager(ctx context.Context, client client.Client, db database.Database, opts LoaderOptions) (*LoaderManager, error) {
+	// initialize struct
+	g, ctx := errgroup.WithContext(ctx)
+	inputCh := make(chan *LoaderMsg[BlockRange])
+	loader := &LoaderManager{
+		client:  client,
+		db:      db,
+		inputCh: inputCh,
+		ctx:     ctx,
+		g:       g,
+	}
+
+	// loaders
+	blockRangeLoader := NewLoader[BlockRange, []*chainhash.Hash](client, inputCh, blockRangeHandler)
+	g.Go(blockRangeLoader.Run)
+	blockHashLoader := NewLoader[[]*chainhash.Hash, []*types.Block](client, blockRangeLoader.Dst(), blockHashHandler)
+	g.Go(blockHashLoader.Run)
+	blockLoader := NewLoaderSink[[]*types.Block](blockHashLoader.Dst(), blockHandler)
+	g.Go(blockLoader.Run)
```

Comment:
btw I found out that you don't have to declare types in a generic function if Go can infer types from args passed in. So you can prob remove the type declarations here.

Reply:
I had left them for the time being because I thought it was kinda nice to have them explicit, but now I'm thinking it'll prob be cleaner without.